File: CocoaSplit/CSAnimationRunner/CSAnimationBlock.py
import objc
import PyObjCTools.AppHelper
from Foundation import NSObject,NSLog
from Quartz import CACurrentMediaTime,CATransaction
from CSAnimation import *



def current_frame():
  return CATransaction.valueForKey_("__CS_BLOCK_OBJECT__")

class CSAnimationDelegate(NSObject):

    @objc.signature('v@:@c')
    def animationDidStop_finished_(self, animation, finished):
        cs_anim = animation.valueForKeyPath_("__CS_COMPLETION__")
        cs_anim.completed()


class AnimationBlock:
    def __init__(self, duration = 0.0):
        cframe = current_frame()
        if cframe:
            self.layout = cframe.layout
            self.animation_module = cframe.animation_module
            self.current_begin_time = cframe.current_begin_time
            self.current_end_time = cframe.current_end_time
        else:
            self.layout = None
            self.animation_module = None
            self.current_begin_time = None
            self.current_end_time = None
        
        CATransaction.begin()
        CATransaction.setValue_forKey_(self, "__CS_BLOCK_OBJECT__")
        self.animations = []
        self.duration = duration
        self.max_animation_time = 0.0
        self.beginTime =  0.0
        self.real_completion_block = None
        self.input_map = {}
        self.label_map = {}
        
        if not self.duration:
            self.duration = 0.25


    def internal_completion_block(self, real_completion):
        if real_completion:
            real_completion()

    def set_completion_block(self, completion_callable):
        if completion_callable:
            CATransaction.setCompletionBlock_(lambda: self.internal_completion_block(completion_callable))


    def add_waitmarker(self, duration=0, target=None, wait_only=None, **kwargs):
        if not self.current_begin_time:
            self.current_begin_time = self.layout.rootLayer().convertTime_fromLayer_(CACurrentMediaTime(), None)

        new_mark = CSAnimation(None, "__CS_WAIT_MARK", None, **kwargs)
        new_mark.isWaitMark = True
        new_mark.duration = duration
        new_mark.cs_input = target
        if new_mark.cs_input and new_mark.cs_input in self.input_map:
            self.current_begin_time = self.input_map[new_mark.cs_input]
        elif new_mark.label and new_mark.label in self.label_map:
            self.current_begin_time = self.label_map[new_mark.label].end_time
        else:
            if not wait_only:
                self.current_begin_time = self.latest_end_time
        new_mark.apply(self.current_begin_time)
        self.latest_end_time = max(self.latest_end_time, new_mark.end_time)
        self.current_begin_time += new_mark.duration
        #self.animations.append(new_mark)
        return new_mark

    def add_animation(self, animation, target, keyPath):
        
        if not self.current_begin_time:
            self.current_begin_time = self.layout.rootLayer().convertTime_fromLayer_(CACurrentMediaTime(), None)
        if animation.duration == 0:
            #hax
            animation.duration = 0.001
            animation.animation.setDuration_(0.001)

        if animation.label:
            self.label_map[animation.label] = animation
        
        NSLog("CURRENT BEGIN %f", self.current_begin_time)
        
        a_duration = animation.apply(self.current_begin_time)

        self.latest_end_time = animation.end_time
        
        if (animation.uukey and animation.target):
            self.animation_module.all_animations[animation.uukey] = animation.target



        if animation.cs_input:
            if animation.cs_input in self.input_map:
                self.input_map[animation.cs_input] = max(self.latest_end_time, self.input_map[animation.cs_input])
            else:
                self.input_map[animation.cs_input] = self.latest_end_time


        if not animation.ignore_wait and animation.animation:
            animation.animation.setValue_forKeyPath_(animation, "__CS_COMPLETION__")
            animation.animation.setDelegate_(CSAnimationDelegate.alloc().init())
        self.animations.append(animation)

        return animation

    def wait(self, duration=0, target=None, **kwargs):
        waitmark = self.add_waitmarker(duration=duration, target=target, wait_only=1, **kwargs)

    def waitAnimation(self, duration=0, target=None, **kwargs):
        return self.add_waitmarker(duration=duration, target=target, **kwargs)

    def commit(self):
        CATransaction.commit()




def setCompletionBlock(completion_block):
    NSLog("SET BLOCK %@", completion_block)
    current_frame().set_completion_block(completion_block)

def wait(duration=0):
    current_frame().wait(duration, None)

def waitAnimation(duration=0, **kwargs):
    current_frame().waitAnimation(duration=duration, **kwargs)

def animationDuration():
    current_frame().duration

def beginAnimation(duration=0.25):

    new_frame = AnimationBlock(duration)


def commitAnimation():
    current_frame().commit()

File: CocoaSplit/CSAnimationRunner/CSAnimationBlock.js
var block_uuid_map = {};

var CSAnimationBlock = {};

CSAnimationBlock.currentFrame = function() {
    
    var blockUUID = CATransaction.valueForKey("__CS_BLOCK_UUID__");
    return block_uuid_map[blockUUID];
}




function AnimationBlock(duration, inherit_frame) {
    
    
    var self = this;
    
    if (duration === undefined)
    {
        duration = 0.0;
    }
    
    var cframe = CSAnimationBlock.currentFrame();
    

    
    if (cframe)
    {
        self.parent_frame_begin_time = cframe.current_begin_time;
    }
    
    
    if (!inherit_frame || !cframe)
    {
        self.layout = getCurrentLayout();
        self.current_begin_time = null;
        self.latest_end_time = null;
        self.animation_info = {};
        self.animation_info["all_animations"] = {};
        self.isolated = !inherit_frame;
        self.input_map = {};
        self.label_map = {};


    } else {
        cframe.applyPendingAnimations();
        self.layout = cframe.layout;
        self.current_begin_time = cframe.current_begin_time;
        self.latest_end_time = cframe.latest_end_time;
        self.animation_info = cframe.animation_info;
        self.isolated = !inherit_frame;
        self.input_map = cframe.input_map;
        self.label_map = cframe.label_map;
    }
    
    self.animations = [];
    self.duration = duration;
    self.max_animation_time = 0.0;
    self.beginTime = 0.0;
    self.real_completion_block = null;
    self.uuid = generateUUID();
    
    if (self.duration == 0.0)
    {
        self.duration = 0.25;
    }
    
    this.internal_completion_block = function(real_completion) {
        if (real_completion)
        {
            real_completion();
        }
    }
    
    this.set_completion_block = function(completion_callable) {
        if (completion_callable)
        {
            
            CATransaction.setCompletionBlockJS(completion_callable);
        }
    }
    
    
    this.advance_begin_time = function(duration) {
        if (self.current_begin_time)
        {
            self.latest_end_time = self.current_begin_time + duration;
        }
    }
    
    
    this.add_animation = function(animation)  {
        
        self.applyPendingAnimations();
        self.animations.push(animation);
    }
    
    
    this.add_animation_real = function(animation) {
        if (!self.current_begin_time)
        {
            
            self.current_begin_time = self.layout.rootLayer.convertTimeFromLayer(CACurrentMediaTime(), null);
        }
        
        if (!self.latest_end_time)
        {
            self.latest_end_time = self.current_begin_time;
        }
        
        if (animation.duration == 0)
        {
            animation.duration = 0.001;
            animation.animation.duration = 0.001;
        }
        
        if (animation.label)
        {
            self.label_map[animation.label] = animation;
        }
        
        if (!animation.ignore_wait && animation.animation)
        {
            var a_delegate = new CSJSAnimationDelegate(animation);
           animation.animation.delegate = a_delegate;
        }
        
        var use_begin_time = self.current_begin_time;
        
        if (animation.cs_input && self.input_map[animation.cs_input.uuid])
        {
            use_begin_time = self.input_map[animation.cs_input.uuid].begin_time;
        }
        
        var a_duration = animation.apply(use_begin_time);
        self.latest_end_time = Math.max(self.latest_end_time, animation.end_time);
        
        if (animation.uukey && animation.target)
        {
            self.animation_info.all_animations[animation.uukey]= animation.target;
        }
        
        if (animation.cs_input)
        {
            if (self.input_map[animation.cs_input.uuid])
            {
                
                self.input_map[animation.cs_input.uuid].end_time = Math.max(animation.end_time, self.input_map[animation.cs_input.uuid].end_time);
                
            } else {
                self.input_map[animation.cs_input.uuid] = {begin_time: self.current_begin_time, end_time: animation.end_time};
            }
        }
        
        //this.animations.push(animation);
        return animation;
    }
    
    
    this.add_waitmarker = function(duration, target, wait_only) {
        if (duration === undefined)
        {
            duration = 0.0;
        }
        new_mark = new CSAnimation(null, "__CS_WAIT_MARK", null);
        new_mark.isWaitMark = true;
        new_mark.duration = duration;
        new_mark.cs_input = target;
        new_mark.wait_only = wait_only;
        self.animations.push(new_mark);
        self.applyPendingAnimations();

        return new_mark;

    }
    
    
    this.add_waitmarker_real = function(waitMark) {
        
        if (!self.current_begin_time)
        {
            self.current_begin_time = self.layout.rootLayer.convertTimeFromLayer(CACurrentMediaTime(), null);
        }
        
        if (!self.latest_end_time)
        {
            self.latest_end_time = self.current_begin_time;
        }
        

        if (waitMark.cs_input)
        {
            var input_begin_time = self.current_begin_time;
            var input_end_time = self.current_begin_time;
            if (self.input_map[waitMark.cs_input.uuid])
            {
                input_begin_time = self.input_map[waitMark.cs_input.uuid].end_time;
                input_end_time = self.input_map[waitMark.cs_input.uuid].end_time;
            }
            input_begin_time += waitMark.duration;
            
            self.input_map[waitMark.cs_input.uuid] = {begin_time: input_begin_time, end_time: Math.max(input_end_time, input_begin_time)}
            
            
            this.latest_end_time = Math.max(self.latest_end_time, input_begin_time);
            return waitMark;
        } else if (waitMark.label && self.label_map[waitMark.label]) {
            self.current_begin_time = self.label_map[waitMark.label].end_time;
        } else {
            if (!waitMark.wait_only)
            {
                self.current_begin_time = self.latest_end_time;
            }
        }
        
        
        waitMark.apply(self.current_begin_time);
        this.latest_end_time = Math.max(self.latest_end_time, waitMark.end_time);
        
        self.current_begin_time += waitMark.duration;
    
        return waitMark;
    }
    
    
    this.applyPendingAnimations = function()
    {
        for(var i = 0, len = self.animations.length; i < len; i++)
        {
            var anim = self.animations[i];
            if (anim.isWaitMark)
            {
                self.add_waitmarker_real(anim);
            } else {
                self.add_animation_real(anim);
            }
        }
        
        self.animations = [];
    }
    
    
    
    this.commit = function()
    {
        
        self.applyPendingAnimations();
        
        CATransaction.setValueForKey(null, "__CS_BLOCK_UUID__");
        self.animation_info = null;
        delete block_uuid_map[self.uuid];
        if (self.doScriptWait)
        {
            self.waitAnimation();
        }
        CATransaction.commit();
        if (!self.isolated)
        {
            let cframe = CSAnimationBlock.currentFrame();
            if (cframe)
            {

                cframe.current_begin_time = self.current_begin_time;
                cframe.latest_end_time = self.latest_end_time;
            }
        }
        

    }
    
    this.waitTransition = function() {
        if (self.parent_frame_begin_time)
        {
            self.current_begin_time = self.parent_frame_begin_time;
        }
    }
    
    
    this.waitAnimation = function(duration, target) {
        return self.add_waitmarker(duration, target, 0);
    }
    
    this.wait = function(duration, target) {
        return self.add_waitmarker(duration, target, 1);
    }
    
    CATransaction.begin();
    self.doScriptWait = false;
    block_uuid_map[self.uuid] = self;
    CATransaction.setValueForKey(self.uuid, "__CS_BLOCK_UUID__");
    //this.current_begin_time = 0;


}

var setCompletionBlock = function(completion_block) {
    

    CSAnimationBlock.currentFrame().set_completion_block(completion_block);
}

var wait = function(duration)
{
    return CSAnimationBlock.currentFrame().wait(duration);
}

var waitAnimation = function(duration) {
    return CSAnimationBlock.currentFrame().waitAnimation(duration);
}

var animationDuration = function() { CSAnimationBlock.currentFrame().duration; }

var beginIsolatedAnimation = function(duration) {
    if (duration === undefined)
    {
        duration = 0.25;
    }
    
    return new AnimationBlock(duration, false);
}


var beginAnimation = function(duration) {
    
    if (duration === undefined)
    {
        duration = 0.25;
    }
    
    return new AnimationBlock(duration, true);
}

var commitAnimation = function() {
    var cframe = CSAnimationBlock.currentFrame();
    cframe.commit();
}

var waitTransition = function () {
    var cframe = CSAnimationBlock.currentFrame();
    if (cframe)
    {
        cframe.waitTransition();
    }
}


var waitScript = function() {
    var cframe = CSAnimationBlock.currentFrame();
    if (cframe)
    {
        cframe.isolated = false;
        cframe.doScriptWait = true;
    }
}
